### PR TITLE
client: close dirfrag whem rmdir

### PR DESCRIPTION
**Pull request by ukernel**
Fixes: #10387
Signed-off-by: Yan, Zheng <zyan@redhat.com>

### CODE REVIEW

**Comment by jcsp** [comment]
This doesn't appear to work: my test case is still seeing the client holding caps after rmdir.

**Comment by ukernel** [comment]
updated, please test. the old code does not work because unsafe request holds reference to dentry, which prevent dirfrag from closing

File: src/client/Client.cc [hunk 1]
```diff
@@ -4145,8 +4145,24 @@ void Client::_schedule_invalidate_dentry_callback(Dentry *dn, bool del)
...
+
+  if (in->dir && !in->dir->dentry_list.empty()) {
+    for (xlist<Dentry*>::iterator p = in->dir->dentry_list.begin();
+	!p.end(); ) {
+      Dentry *dn = *p;
+      ++p;
+      if (dn->lru_is_expireable())
+	unlink(dn, false, false);  // close dir, drop dentry
+    }
+    --ref;
```
Comment (jcsp) [inline src/client/Client.cc]: Is it correct that `ref` is decremented only once, no matter how many dentries we unlinked?

**Comment by jcsp** [comment]
This passes the test now

File: src/client/Client.cc (hunk 1 above)
Comment (ukernel) [inline src/client/Client.cc]: yes. opened dir holds one reference. dentries do not hold reference.

**Comment by gregsfortytwo** [comment]
This sounds okay to me, but please review properly and merge when you're happy @jcsp. :)

**Comment by jcsp** [comment]
I'm happy with this, merging

**Pull request merged**